### DISCUSSION TITLE
how to make lsp amazing

### DISCUSSION THREAD

**Discussion by sdancer** [gleam-lang/gleam#2600] [General]
to quote Louis "productivity isn't just about having a good enough language", of course we have that, gleam is becoming great... "but also about having good enough tooling", lsp is one of those tools that make life of new and/or experienced coders easy.. with that said:

prior art on incrementalized computation:
- https://www.youtube.com/watch?v=N6b44kMS6OM (rust)
- https://github.com/salsa-rs/salsa
- (fungi) https://arxiv.org/abs/1808.07826

stuff to do:
- extract the function signatures on all the modules where parse success, even if said functions body don't properly infer, and add/use in the lsp context
- cache functions, even if we break the file parsing by typing while missing a closing bracket, still get helpful autocompletion/etc

user experience behaviors that would be desirable to have:
=
- focus on the file open first (comes back to the lsp as a partial query-response engine)

metadata storage
=
- replace Module structure serialization with fully macro derived serialization to preserve most or all the data post inference, using serde bson or similar, whatever yields a good balance between load performance and easiness of usage

**Comment by lpil** [comment]
How long is the delay before you get results? So far the complaints I've got is that the LSP is too fast.

> add incremental compilation
> cache modules (albeit the unit of work should be function rather than module)

We do both of these already. 👍 

> cache functions, invalidate the function dependency graph on changes and recompile only affected ones

With how fast Gleam compiles I think it'd need to be a very large module with very large functions for this to be beneficial.

**Reply by lpil** [reply]
How long is it taking you to get results?

So far people have been complaining that it is too fast and have asked me to add artificial delays to compilation as the rate at which diagnostics are presented is reportedly distracting.

**Reply by lpil** [reply]
I'm not sure what you're showing there. Each time in the video the diagnostic was shown within fractions of a second?

**Reply by lpil** [reply]
I'm sorry I don't understand. There are diagnostics on screen in all of the times you've just said there are not.
